### PR TITLE
Start running bazel C/C++ tests on mac

### PR DESCRIPTION
**Pull request by jtattermusch**
Setup continuous C/C++ test under bazel.
The tests are setup very similar to linux/windows RBE tests, with the exception that Foundry doesn't support Mac executors, so the tests are still built and run locally (but the test results are uploaded to result store).

I will enable the tests on PRs once the test results on master are stable.

There is also a number of fixes to the bazel tests on mac:
- added grpc_init() and grpc_shutdown() to some tests, because due to https://github.com/bazelbuild/bazel/issues/4341 we are using GPR_PTHREAD_TLS on mac, which requires initializing thread local variables before using them.
- under bazel, GRPC_CFSTREAM is enabled and the port.h needed to be cleaned up so that the tests continue working
- a few more fixes (see diffs).

### CODE REVIEW

**Comment by jtattermusch** [comment]
Current failures from a local run
https://source.cloud.google.com/results/invocations/986ca05e-37c1-4cda-9686-755949594aea/targets

`bazel --bazelrc=tools/remote_build/mac.bazelrc test //test/... --keep_going`

**Comment by jtattermusch** [comment]
The other problem is that with this hack:
https://github.com/grpc/grpc/blob/a0ad4081cbea0f3e524d4fc51c6a5c7dbbf3d6dc/include/grpc/impl/codegen/port_platform.h#L236, on macos under bazel we're using GPR_PTHREAD_TLS instead of GPR_GCC_TLS, but that actually fixed our macos bazel but it breaks some of the tests
because the thread local storage in uninitialized. E.g.

```
Executing tests from //test/core/backoff:backoff_test
-----------------------------------------------------------------------------
D1017 10:13:21.109471000 4373829056 test_config.cc:384]                test slowdown factor: sanitizer=1, fixture=1, poller=1, total=1
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from BackOffTest
[ RUN      ] BackOffTest.ConstantBackOff
E1017 10:13:21.110358000 4373829056 tls_pthread.cc:26]                 assertion failed: 0 == pthread_setspecific(tls->key, (void*)value)
```

**Comment by jtattermusch** [comment]
Results so far: https://source.cloud.google.com/results/invocations/cf20b310-8718-427c-a5f5-a9d7de5daed6/targets

**Comment by jtattermusch** [comment]
https://source.cloud.google.com/results/invocations/e7938d61-8f53-4f1a-a638-2f7abbcfa6c3

**Comment by jtattermusch** [comment]
https://source.cloud.google.com/results/invocations/bdf17ac6-9bba-4b32-a7fc-21197c7fb935

**Comment by jtattermusch** [comment]
I included the changes from https://github.com/grpc/grpc/pull/20864.

**Comment by jtattermusch** [comment]
https://source.cloud.google.com/results/invocations/7c138a11-2490-4e7e-8523-b35f9106ecb5

**Comment by jtattermusch** [comment]
@veblush  @rmstar  this is ready for review.

An adhoc run with all the tests passing: https://source.cloud.google.com/results/invocations/5c6c4d70-d6a2-491e-adf0-6b9dcb9e7d27/targets   I still might need to look into test flakes, but this is good enough for a start.

File: test/core/security/credentials_test.cc [hunk 1]
```diff
@@ -1440,16 +1440,32 @@ static void test_metadata_plugin_failure(void) {
 static void test_get_well_known_google_credentials_file_path(void) {
   char* path;
   char* home = gpr_getenv("HOME");
+  bool restore_home_env = false;
+#if defined(GRPC_BAZEL_BUILD) && \
+    (defined(GPR_POSIX_ENV) || defined(GPR_LINUX_ENV))
+  // when running under bazel with locally, the HOME variable is not set
```
Comment (rmstar) [inline test/core/security/credentials_test.cc]: nit: bazel with locally -> bazel locally

**Comment by jtattermusch** [comment]
Known failures: 
https://github.com/grpc/grpc/issues/21297
https://github.com/grpc/grpc/issues/20459
https://github.com/grpc/grpc/issues/21309

**Pull request merged**